src/encode.rs

```diff
@@ -0,0 +1,296 @@
+use crate::{
+    errors::{Error, Result},
+    key::PublicKeyParts,
+    parse::rsa_oid,
+    RSAPrivateKey, RSAPublicKey,
+};
+use num_bigint::{BigUint, ToBigInt};
+use num_traits::Zero;
+use simple_asn1::{to_der, ASN1Block};
+
+#[cfg(feature = "pem")]
+impl RSAPrivateKey {
+    /// Converts RSA Private key into `PKCS1` encoded bytes in pem format.
+    ///
+    /// Encodes the key with the header:
+    /// `-----BEGIN RSA PRIVATE KEY-----`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use rsa::RSAPrivateKey;
+    /// use rand::rngs::OsRng;
+    ///
+    /// let mut rng = OsRng;
+    /// let bits = 2048;
+    /// let private_key = RSAPrivateKey::new(&mut rng, bits).expect("failed to generate a key");
+    ///
+    /// let _ = private_key.to_pem_pkcs1();
+    /// ```
+    pub fn to_pem_pkcs1(&self) -> Result<String> {
+        let pem = pem::Pem {
+            tag: String::from("RSA PRIVATE KEY"),
+            contents: self.to_pkcs1()?,
+        };
+        Ok(pem::encode(&pem))
+    }
+
+    /// Converts RSA Private key into `PKCS8` encoded bytes in pem format.
+    ///
+    /// Encodes the key with the header:
+    /// `-----BEGIN PRIVATE KEY-----`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use rsa::RSAPrivateKey;
+    /// use rand::rngs::OsRng;
+    ///
+    /// let mut rng = OsRng;
+    /// let bits = 2048;
+    /// let private_key = RSAPrivateKey::new(&mut rng, bits).expect("failed to generate a key");
+    ///
+    /// let _ = private_key.to_pem_pkcs8();
+    /// ```
+    pub fn to_pem_pkcs8(&self) -> Result<String> {
+        let pem = pem::Pem {
+            tag: String::from("PRIVATE KEY"),
+            contents: self.to_pkcs8()?,
+        };
+        Ok(pem::encode(&pem))
+    }
+}
+
+#[cfg(feature = "pem")]
+impl RSAPublicKey {
```

Comment:
Similarly here: `trait PublicKeyPemEncoding : PublicKeyEncoding` and then auto-implementations.